sequencer/finalizer.go

```diff
@@ -273,11 +296,86 @@ func (f *finalizer) halt(ctx context.Context, err error) {
+	if f.proverID != "" && f.proverID != proverID {
+		event := &event.Event{
+			ReceivedAt:  time.Now(),
+			Source:      event.Source_Node,
+			Component:   event.Component_Sequencer,
+			Level:       event.Level_Critical,
+			EventID:     event.EventID_FinalizerRestart,
+			Description: fmt.Sprintf("proverID changed from %s to %s, restarting sequencer to discard current WIP batch and work with new executor", f.proverID, proverID),
+		}
+
+		err := f.eventLog.LogEvent(context.Background(), event)
+		if err != nil {
+			log.Errorf("error storing payload: %v", err)
+		}
+
+		log.Fatal("restarting sequencer to discard current WIP batch and work with new executor")
+	}
+	f.storedFlushID = storedFlushID
+}
+
+func (f *finalizer) storePendingTransactions(ctx context.Context) {
+	storedFlushID, proverID, err := f.dbManager.GetStoredFlushID(ctx)
+	if err != nil {
+		log.Errorf("failed to get stored flush id, Err: %v", err)
+	} else {
+		f.storedFlushID = storedFlushID
+		f.proverID = proverID
+	}
+	for {
+		txsToStoreCount := len(f.pendingTransactionsToStore)
+		// Save CPU cycles if there is nothing to save
+		if txsToStoreCount == 0 {
+			time.Sleep(100 * time.Millisecond) //nolint:gomnd
+		} else {
+			for txsToStoreCount > 0 && f.pendingTransactionsToStore[0].flushId <= f.storedFlushID {
+				tx := f.pendingTransactionsToStore[0]
+				f.storeProcessedTx(ctx, tx.batchNumber, tx.coinbase, tx.timestamp, tx.oldStateRoot, tx.response, tx.isForcedBatch)
+
+				if tx.txTracker != nil {
+					f.updateWorkerAfterTxStored(ctx, tx.txTracker, tx.batchResponse)
+				}
+
+				f.pendingTransactionsToStoreMux.Lock()
+				txsToStoreCount = len(f.pendingTransactionsToStore)
+				if txsToStoreCount > 1 {
+					f.pendingTransactionsToStore = f.pendingTransactionsToStore[1:]
+				} else {
+					f.pendingTransactionsToStore = []transactionToStore{}
+				}
+				f.pendingTransactionsToStoreMux.Unlock()
+				txsToStoreCount = txsToStoreCount - 1
+				f.pendingTransactionsToStoreWG.Done()
+			}
+
+			if txsToStoreCount > 0 {
+				time.Sleep(100 * time.Millisecond) //nolint:gomnd
+				storedFlushID, proverID, err := f.dbManager.GetStoredFlushID(ctx)
+				if err != nil {
+					log.Errorf("failed to get stored flush id, Err: %v", err)
+				} else {
+					f.checkProverIDAndUpdateStoredFlushID(storedFlushID, proverID)
+				}
+			}
+		}
+	}
+}
+
 // newWIPBatch closes the current batch and opens a new one, potentially processing forced batches between the batch is closed and the resulting new empty batch
 func (f *finalizer) newWIPBatch(ctx context.Context) (*WipBatch, error) {
 	f.sharedResourcesMux.Lock()
 	defer f.sharedResourcesMux.Unlock()
 
+	// Wait until all processed transactions are saved
+	startWait := time.Now()
+	f.pendingTransactionsToStoreWG.Wait()
+	endWait := time.Now()
+
+	log.Info("waiting for pending transactions to be stored took: ", endWait.Sub(startWait).String())
```

Comment:
it will be good to put in the message log the unit of the time (miliseconds(ms)?)